marshal.go

```diff
@@ -0,0 +1,367 @@
+// If "into" is not a struct, this method will fail.
+// As this uses go's reflect package, only exported names can be unmarshalled.
+// Strict mode requires that all fields of the struct are set and defined exactly once.
+// You can set struct tags to influence the unmarshalling process.
+// All tags must have the form `tadl:"..."` and are a list of comma separated identifiers.
+//
+// The first identifier can be used to rename the field, so that an element with the renamed
+// name is parsed, and not the name of the struct field.
+//
+//  // This tadl snippet...
+//  #!{item{...}}
+//  // could be unmarshalled into this go struct.
+//  type Example struct {
+//      SomeName Content `tadl:"item"`
+//  }
+//
+// The second identifier is used to specify what kind of thing is being parsed.
+// This can be used to parse attributes (attr) or text (text).
+//
+// Attributes can be parsed into primitive types: string, bool and the integer (signed & unsigned) and float types.
+// Should the value not be valid for the target type, e.g. an integer that is too large or a negative value for an uint,
+// an error is returned describing the issue.
+//
+//  // This tadl snippet...
+//  #item @key{value} @X{123}
+//  // could be unmarshalled into this go struct.
+//  type Example struct {
+//      SomeName string `tadl:"key,attr"` // Notice how you can rename an attribute
+//      X        int    `tadl:",attr"` // You can choose to not rename it, by omitting the rename parameter.
+//  }
+//
+// Text can be parsed into fields marked with 'text'.
+// In normal mode the field will have all text that occurred in the element concatenated or an empty string if no
+// text was inside the element.
+// In strict mode exactly one text item with length > 0 is expected.
+// Renaming a text parameter is not an error but pointless, so best leave the rename parameter empty.
+//
+func Unmarshal(r io.Reader, into interface{}, strict bool) error {
+	parse := parser.NewParser("", r)
+
+	if into == nil {
+		return fmt.Errorf("cannot unmarshal into nil")
+	}
+
+	tree, err := parse.Parse()
+	if err != nil {
+		return NewUnmarshalError(tree, "parser error", err)
+	}
+
+	value := reflect.ValueOf(into)
+	unmarshal := unmarshaler{strict: strict}
+
+	if err := unmarshal.node(tree, value); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// unmarshaler is a helper struct for easier managing the unmarshalling process.
+type unmarshaler struct {
+	strict bool
+}
+
+// While unmarshalling we might need to process a node as an attribute.
+// We use this enum to make the decision.
+type unmarshalType int
+
+const (
+	unmarshalNormal unmarshalType = iota
+	unmarshalAttribute
+	unmarshalText
+)
+
+// UnmarshalError is an error that occurred during unmarshaling.
+// It contains the offending node, a string with details and an underlying error (if any).
+type UnmarshalError struct {
+	node     *parser.TreeNode
+	detail   string
```

Comment:
Position für Rust-Like Fehler? Public-Methoden?

Reply:
Die Positionen sind alle in der node gespeichert, aber die sollte auch public sein, das stimmt.